### DISCUSSION TITLE
Using RTP library

### DISCUSSION THREAD

**Discussion by tomsasson-fibernet** [ireader/media-server#200] [Q&A] [answered]
Hey @ireader,
Great project! I was trying to use RTP library, but cannot do that.
Could you post a test or example for using RTP in RFC3550 form and with MPEGTS?

**Comment by ireader** [comment, marked as answer]
1. At librtsp/source/utils,  rtsp-muxer & rtsp-demuxer support mpeg-ts rtp packet "encode/decode"
2. At librtp/test， rtp-receiver-test is a example for rtp client receive udp & tcp packet.
3. At librtsp/test & librtsp/test/media,  have a rtsp-server example, support raw h264 & ps & mp4 file vod play.

**Reply by tomsasson-fibernet** [reply]
Amazing! Thanks!

**Reply by tomsasson-fibernet** [reply]
Hi,
I still have an issue with MPEGTS packets over RTP. My wireshark doesn't decode the packets the way I would expect.
What I did is I use rtp_sender_test, but I am not sure that simply defining "packet" array is enough, there seems to be no RTP header in this test.
This function receives "rtpfile", now what I wonder is how do I create this file from the first place in MPEGTS packets wrapped by RTP packet.

**Comment by tomsasson-fibernet** [comment]
Hi @ireader ,
I still have an issue with MPEGTS packets over RTP. My wireshark doesn't decode the packets the way I would expect.
What I did is I use rtp_sender_test, but I am not sure that simply defining "packet" array is enough, there seems to be no RTP header in this test.
This function receives "rtpfile", now what I wonder is how do I create this file from the first place in MPEGTS packets wrapped by RTP packet.

What I want is to use "RTP with MPEG-2 PS/TS" as depicted in the Readme file, but too bad there are no comments in the code so I don't know where to approach.

By the way, later on if I succeed with my project, I will be able to help and comment the code so it will be more user-friendly, or add tests

**Reply by ireader** [reply]
"rtp_sender_test" load local rtp binary file and send to target host:port.

The File Format:
two-bytes rtp packet length(network byte order, big-endian) + rtp packet

e.g.   A mpeg-ts rtp packet length with 200 (12-RTP-HEADER+188-TS)
00 C8 RTP-PACKET 00 C8 RTP-PACKET 00 C8 RTP-PACKET ....

**Reply by ireader** [reply]
If you have a mpeg-ts file, and want send with rtp protocol:
1. rtp_payload_encode_create(RTP_PAYLOAD_MP2T, "MP2T", seq, ssrc, ...)    (see more rtp-payload-test.cpp)
2. load mpeg-ts packet (188-bytes)
3. rtp_payload_encode_input(rtp, mpeg-ts-packet-188-bytes, 188, mpeg-ts-packet-timestamp)
4. loop 2/3 until file end
5. on rtp_payload_encode_create hander->packet callback do send
6. rtp_payload_encode_destroy(rtp)

**Reply by ireader** [reply]
librtsp/source/utils/rtsp-muxer.c can use to packet video/audio packet:
1. int pid = rtsp_muxer_add_payload(muxer, "RTP/AVP",  90000, RTP_PAYLOAD_MP2T, "MP2T", seq, ssrc, port, NULL, 0);
2. int video = rtsp_muxer_add_media(muxer, pid, RTP_PAYLOAD_H264, NULL, 0);  // h264
3. int audio = rtsp_muxer_add_media(muxer, pid, RTP_PAYLOAD_MP4A, NULL, 0); // aac
4. rtsp_muxer_input(muxer, video, pts, dts, data, bytes, KEYFRAME); // write video

**Reply by Znroee** [reply]
Hi ireader,
Thanks for your provided info about using the RTP library against local file.
but, unfortunately, our video source is not a binary file, it is a stream pipe of real time video over UDP connection.

please tell us, what is the procedure (functions, steps) to wrap our video stream in RTP format. (in other word: take every raw data video frame that received and translate it to RTP packet).

Thanks a`lot.
Roee.

**Reply by ireader** [reply]
Which kind live stream in your scene?

For H.264 live video stream,  here it is:
> 
```c
int pid = rtsp_muxer_add_payload(muxer, "RTP/AVP", 90000, RTP_PAYLOAD_H264, "H264", <seq>, <ssrc>, <port>, NULL, 0);
int video = rtsp_muxer_add_media(muxer, pid, RTP_PAYLOAD_H264, <avc decoder configuration record>, <bytes>); // h264

while(true) {
     // for each h264 frame
     rtsp_muxer_input(muxer, video, pts, dts, data, bytes, KEYFRAME); // write video
}
```

**Reply by Znroee** [reply]
Hi,

Thank you for your rapid response,
"rtsp_muxer_input" seems like a part from RTSP, while we need support in RTP (over UDP). 
it is o.k to use this method also ?

**Reply by ireader** [reply]
Yes, it's RTP muxer.
